### PR TITLE
ARROW-6883: [C++][Python] Allow writing dictionary deltas

### PR DESCRIPTION
**Pull request by pitrou**
* Add an ipc::IpcWriteOptions member to govern emission of dictionary deltas.

  If the option is enabled, deltas are detected by checking whether the new
  dictionary starts with the last emitted one for the same field.
  However, for nested dictionaries, deltas are not emitted for the outer dictionary,
  as the read path doesn't support it.

* Add a stats() method to ipc::StreamDecoder

* Expose the IPC statistics in Python, and add tests

### CODE REVIEW

**Comment by wesm** [comment]
Acknowledged, will review when I can

**Comment by pitrou** [comment]
TODO: need to update the feature matrix

**Comment by pitrou** [comment]
cc @lidavidm

**Comment by lidavidm** [comment]
The Flight changes look good to me (though there's still ARROW-10787).

**Review by wesm — APPROVED** [review]
+1 here. Sorry for the long delay

File: cpp/src/arrow/ipc/options.h [hunk 1]
```diff
@@ -65,6 +65,20 @@ struct ARROW_EXPORT IpcWriteOptions {
...
 
+  /// \brief Whether to emit dictionary deltas
+  ///
+  /// If false, a changed dictionary for a given field will emit a full
+  /// dictionary replacement.
+  /// If true, a changed dictionary will be compared against the previous
+  /// version. If possible, a dictionary delta will be omitted, otherwise
+  /// a full dictionary replacement.
+  ///
+  /// Default is false to maximize stream compatibility.
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/ipc/options.h:76]: No, I simply meant that it's off for compatibility with implementations that don't support delta dictionaries.

**Pull request closed without merging**